### PR TITLE
[build] Consolidate stdlib's .v files under a single directory.

### PR DESCRIPTION
**Pull request by ejgallego**
Currently, `.v` under the `Coq.` prefix are found in both `theories`
and `plugins`. Usually these two directories are merged by special
loadpath code that allows double-binding of the prefix.

This adds some complexity to the build and loadpath system; and in
particular, it prevents from handling the `Coq.*` prefix in the
simple, `-R theories Coq` standard way.

We thus move all `.v` files to theories, leaving `plugins` as an
OCaml-only directory, and modify accordingly the loadpath / build
infrastructure.

Note that in general `plugins/foo/Foo.v` was not self-contained, in
the sense that it depended on files in `theories` and files in
`theories` depended on it; moreover, Coq saw all these files as
belonging to the same namespace so it didn't really care where they
lived.

This could also imply a performance gain as we now effectively
traverse less directories when locating a library.

See also discussion in #10003 , which this PR supersedes.

### CODE REVIEW

**Comment by ejgallego** [comment]
Submitting as a draft PR but it should be ready to merge; assignee needed.

File: Makefile.doc
Comment (Zimmi48) [inline Makefile.doc]: If you rename the `stdlib` target, you should rename `full-stdlib` as well. But is this really a good idea to rename these targets (may break some existing packaging scripts) given that this legacy Makefile is going away in favor of Dune soon?

Comment (ejgallego) [reply Makefile.doc]: Thanks for the review @Zimmi48 , I kept this particular renaming as the target did interact badly with having a `stdlib` directory which I thought could happen at some point in the future.

I dunno indeed; any of the choices [fixing the doc or reverting the target renaming] is fine to me.

Comment (Zimmi48) [reply Makefile.doc]: OK I get it, thanks! Renaming makes sense in this case. I think I would prefer `stdlib-doc` (sounds closer to English) but not a big deal.

Comment (ejgallego) [reply Makefile.doc]: Doc fixed, thanks; I've kept the `doc-` prefix as there are some other `doc-` targets already available.

**Comment by ejgallego** [comment]
I think @coq/build-maintainers may be the right assignee for this one.

File: .github/CODEOWNERS
Comment (Zimmi48) [inline .github/CODEOWNERS]: Why no `theories/nsatz/`? This is not empty.

File: .github/CODEOWNERS
Comment (Zimmi48) [inline .github/CODEOWNERS]: Same question for `theories/setoid_ring/`.

File: toplevel/coqinit.ml
Comment (Zimmi48) [inline toplevel/coqinit.ml]: Why keep the `with_ml` option of the `build_stdlib_path` since it's not used anymore?

File: .github/CODEOWNERS
Comment (ejgallego) [reply .github/CODEOWNERS]: Good catch, thanks.

File: .github/CODEOWNERS
Comment (ejgallego) [reply .github/CODEOWNERS]: Fixed, thanks.

File: toplevel/coqinit.ml
Comment (ejgallego) [reply toplevel/coqinit.ml]: Well seen thanks, fixed.

**Review by Zimmi48 — APPROVED** [review]
This looks fully ready to me.

**Comment by Zimmi48** [comment]
@coq/core: I intend to merge this next Tuesday if no more comments by then. I suggest that we mark this PR as "ready for review" a few hours before merging so that plugins maintainers are notified of this change, even if we don't actually need their review.

**Comment by herbelin** [comment]
Can someone make a brief summary of the pros and cons of the PR?

**Comment by ejgallego** [comment]
@herbelin

The main "pro" this organization brings is that now we can use a simpler rule for binding the `Coq` préfix , basically everything that's under `theories` as compared to now where we merge two directories . This simplies quite a bit the building with dune as a hybrid setup where ml and v files are mixed would require some extra dune files due to the ml and v scoping rules being a bit different.

In general I'm leaning towards a separation of the ml / v parts as for example we'd like to install ml plugins in the standard `lib/ocaml` so they can be seen by the regular ocaml toolchain.

In terms of cons, other than the moving of code there should be none.

**Comment by Zimmi48** [comment]
Note that I forgot to write that down for the record, but this PR (or its previous version) was brought up and discussed in one of the recent calls. A few things changed compared to the previous version following the feedback that was given during this call, including not moving any of the ML files, and not creating a `stdlib` directory.

**Comment by herbelin** [comment]
> A few things changed compared to the previous version following the feedback that was given during this call, including not moving any of the ML files, and not creating a stdlib directory.

OK, thanks.

> In general I'm leaning towards a separation of the ml / v parts as for example we'd like to install ml plugins in the standard lib/ocaml so they can be seen by the regular ocaml toolchain.

OK.

Two related comments:

- From some point of view, `.v` and `.ml` files have interleaved dependencies. For instance, `newring.ml` morally depending on `PEadd` in `Ring_polynom` (via the `plugins.setoid_ring.add` registering). Couldn't we imagine a long-term world where `ocamlc` has a way to checking statically that all `.v` constants the `.ml` code of the plugin needs are present?

- I'm actually not fully satisfied with the existing organization of files. For instance, I'm not sure that I find very intuitive that say, `Coq.btauto` or `Coq.extraction` is under the same namespace where we can find `Coq.Lists` and `Coq.Arith`, compared to, say, having them in `Coq.Tactics.btauto` (or even `Coq.Tactics.Btauto`) or `Coq.Tools.Extraction`. I'm unsure also that `Coq.ssr.ssrbool` (library contents) and `Coq.ssr.ssreflect` (features) have to be in the same namespace. I would intuitively see `ssreflect.v` in `Coq.Tactics.Ssreflect` and the contents of `ssrfun.v` and `ssrbool.v` split and valorized in a sequence of files with suggestive names. For instance, the definitions of the section `OperationProperties` of `ssrfun.v` could go in a namespace say `Coq.Properties`, and the properties of relations defined in `Relation_Definitions.v` could go as well in this namespace. Also, there are a lot of things in `ssrbool.v`. I'm not very familiar with its contents and an `ssrbool` expert would explain better, but I imagine that the part of `ssrbool.v` which is specifically about the type `bool` is presumably expected to be in `Coq.Bool` from the point of view of any reasonable user. (In passing, I do not necessarily expect that a given content has to be in a unique namespace - the same way as a research work can be classified under several topics -, but I expect that users find easily what they are looking for in namespaces which are intuitive for them).

This being said, I'm ok with this technical PR which is apparently useful for Dune.

What I'm hoping is that stdlib2 is going to address the kind of issues I'm mentioning above. I also believe that techniques used in validating software development techniques by means of interviews and statistics should be used for stdlib2 to validate that stdlib2 correctly answers to the need of users.

**Comment by ejgallego** [comment]
> * From some point of view, `.v` and `.ml` files have interleaved dependencies. For instance, `newring.ml` morally depending on `PEadd` in `Ring_polynom` (via the `plugins.setoid_ring.add` registering). Couldn't we imagine a long-term world where `ocamlc` has a way to checking statically that all `.v` constants the `.ml` code of the plugin needs are present?

That seems pretty tricky to do; you'd need to be able to check types are compatible, etc... Also plugins may be used by different `.v` files; I don't really have an idea on how that would work. Maybe the closest we could do is to have some ML quoting inside of `.v` files.

> * I'm actually not fully satisfied with the existing organization of files. For instance, I'm not sure that I find very intuitive that say, `Coq.btauto` or `Coq.extraction` is under the same namespace where we can find `Coq.Lists` and `Coq.Arith`, [...]

Yup, indeed it is a bit surprising at first for plugins to be under `Coq.` , I guess that PR makes the situation more explicit. IIANM this PR is 100% orthogonal to a possible reorganization of the stdlib.

**Comment by herbelin** [comment]
> you'd need to be able to check types are compatible, etc

OCaml remains here a metalanguage so, checking types would be costly. This is a general question for metalanguages actually.

> Also plugins may be used by different .v files

That should not be a problem.

>  IIANM this PR is 100% orthogonal to a possible reorganization of the stdlib.

Yes, let's go with the PR first.

**Comment by ejgallego** [comment]
@herbelin a more general question is whether we'd like to deprecate / remove the current capability that the loadpath code has to merge/rebind prefixes. What do you think?

I have mixed feelings about it; so maybe indeed we should start a bit of discussion on the "coq library model 2.0" and see how we'd like namespace binding to work [I had a bit of discussion with @Zimmi48 about this]

For example I think that we don't want to install every package "globally" under `user-contrib`, but have a more refined hiearchy that the build system puts in scope, etc...

We will see, I think that from my part I may try to start a discussion once the basic steps on the remodelling of the build system are completed; for example I deem crucial to merge the implementations of loadpath in `coq/coqdep` first [this will be doable once we get rid of the coqdep_boot step and in general of the makefiles.

**Comment by herbelin** [comment]
I don't know what you mean by capability that the loadpath code has to merge/rebind prefixes, but thinking at the name space in terms of a "universal" library with a refined hierarchy seems good.

**Comment by ejgallego** [comment]
> I don't know what you mean by capability that the loadpath code has to merge/rebind prefixes, but thinking at the name space in terms of a "universal" library with a refined hierarchy seems good.

I mean that currently we can do `-R dir1 Coq -R dir2 Coq` and both `dir1`, `dir2` are merged.

**Comment by Zimmi48** [comment]
>From some point of view, `.v` and `.ml` files have interleaved dependencies. For instance, `newring.ml` morally depending on `PEadd` in `Ring_polynom` (via the `plugins.setoid_ring.add` registering).

These dependencies to constants being registered in `.v` files do not create actual dependencies to these particular `.v` files. This is what allows stdlib2 to provide alternative `.v` files where the registering is done and to still reuse the ML parts. Trying to verify that these `.v` files are present when compiling would restrict the reusability of these ML pieces.

>I'm actually not fully satisfied with the existing organization of files. For instance, I'm not sure that I find very intuitive that say, `Coq.btauto` or `Coq.extraction` is under the same namespace where we can find `Coq.Lists` and `Coq.Arith`, compared to, say, having them in `Coq.Tactics.btauto` (or even `Coq.Tactics.Btauto`) or `Coq.Tools.Extraction`.

We could even consider splitting some of these out of the `Coq` prefix, although this could create circular dependencies at the library level.

>In passing, I do not necessarily expect that a given content has to be in a unique namespace - the same way as a research work can be classified under several topics

I think we should not confuse namespace and topic.

>I expect that users find easily what they are looking for in namespaces which are intuitive for them

Expecting users to have some intuition of where to look for a given definition looks wrong to me. We should rather improve search features so that this doesn't have to happen.

>IIANM this PR is 100% orthogonal to a possible reorganization of the stdlib.

Indeed, so maybe let's split the discussion on stdlib reorganization elsewhere and avoid pushing too many comments in this specific thread.

**Comment by herbelin** [comment]
@ejgallego:

> I mean that currently we can do -R dir1 Coq -R dir2 Coq and both dir1, dir2 are merged.

I see, this is indeed a problem.

@zimmi48:
> Trying to verify that these .v files are present when compiling would restrict the reusability of these ML pieces.

In this sense, yes, you're right.

> I think we should not confuse namespace and topic.

Namespace can be used for other classifications than topics (e.g. by author of the library?) but using it for topics seem good too (knowing that there is no need to have a given object or concept available only through a unique name). Or am I missing what you have in mind?

> Expecting users to have some intuition of where to look for a given definition looks wrong to me. 

Why? It seems very common for users, when e.g. looking for a lemma on `list` to look at https://coq.github.io/doc/master/stdlib/Coq.Lists.List.html.

> We should rather improve search features so that this doesn't have to happen.

In any case, yes!

> Indeed, so maybe let's split the discussion on stdlib reorganization elsewhere and avoid pushing too many comments in this specific thread.

Yes.

**Comment by ejgallego** [comment]
> I see, this is indeed a problem.

I wouldn't say that a problem _per se_ but more of a question in terms of what semantics we'd like to have. Maybe there are very interesting use cases for this kind of layout; personally, in all senses related to library organization I feel I am pretty open to ideas.

This PR however fits IMO more in the lines of "_prendre du recul pour prendre de l'élan_" , that is to say, we are trying to simplify and cleanup the way libs are handled today, so we can start from a relatively clean model and provide more advanced features that users undoubtedly need.

**Comment by Zimmi48** [comment]
Coming back to the actual topic of the PR, I am proceeding with the announced plan by marking this PR as ready for review and intend to merge it in a few hours. Please, let's add no more off-topic comments (and move this very interesting discussion to another thread).

**Comment by Zimmi48** [comment]
This PR is a purely internal clean-up (no user-visible change) creating a clear separation for plugins that are part of the Coq standard library between their OCaml code (which stays where it was in `plugins/<plugin-name>`) and their Coq code (which is moved to `theories/<plugin-name>`), motivated in particular by simplifying the build process and the Dune integration.

This change concerns all stdlib plugin authors / maintainers as this means that some of your code has moved. However, we do not actually need all the requested reviews (this is more a way of advertising the change).

Following previous discussion in this thread, in #10003 and in a Coq Call, I intend to merge this PR in a few hours if no more comments by then.

**Comment by ejgallego** [comment]
Thanks @Zimmi48

**Pull request merged**
